lib/ui/painting.dart

```diff
@@ -1342,48 +1336,23 @@ class Paint {
   ///
   /// When a shape is being drawn, [colorFilter] overrides [color] and [shader].
   ColorFilter get colorFilter {
-    switch (_data.getInt32(_kColorFilterOffset, _kFakeHostEndian)) {
-      case ColorFilter._TypeNone:
-        return null;
-      case ColorFilter._TypeMode:
-        return ColorFilter.mode(
-          Color(_data.getInt32(_kColorFilterColorOffset, _kFakeHostEndian)),
-          BlendMode.values[_data.getInt32(_kColorFilterBlendModeOffset, _kFakeHostEndian)],
-        );
-      case ColorFilter._TypeMatrix:
-        return ColorFilter.matrix(_objects[_kColorFilterMatrixIndex]);
-      case ColorFilter._TypeLinearToSrgbGamma:
-        return const ColorFilter.linearToSrgbGamma();
-      case ColorFilter._TypeSrgbToLinearGamma:
-        return const ColorFilter.srgbToLinearGamma();
+    if (_objects == null || _objects[_kColorFilterIndex] == null) {
+      return null;
     }
-
-    return null;
+    return _objects[_kColorFilterIndex].creator;
   }
 
   set colorFilter(ColorFilter value) {
```

Comment:
I like the overall approach! My only concern is the test coverage. We could have a C++ unit test to cover paint.cc.

It might be a little more challenging to add a unit test for the Dart code.  I wonder if we can use the fixture similar to the `shell_test.dart` that used in https://github.com/flutter/engine/blob/master/shell/common/shell_unittests.cc#L156. It allows the C++ unit test to specify a Dart function as the entry point, and a C++ native callback. I've used it to set a `onReportTimings` callback on the Dart side, and check its results on the C++ side (https://github.com/flutter/engine/blob/master/shell/common/shell_unittests.cc#L315).

I think it would be nice if we can create the `ColorFilter` on the Dart side, and check the `SkColorFilter` object is properly set on the C++ side. CC @chinmaygarde who might have more suggestions how to do this kind of Dart fixture test.

Reply:
I've added dart tests that will serve to test the C++ side, and that cover the same code.  LMK if you think we need more.

Reply:
https://github.com/flutter/engine/pull/9668/commits/8935830cfb898c9e9fa7ac1613bc2f63275c5de9